pkg/ingester/ingester_profiling.go

```diff
@@ -0,0 +1,345 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ingester
+
+import (
+	"context"
+	"net/http"
+	"runtime/pprof"
+
+	"github.com/grafana/dskit/tenant"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/grafana/mimir/pkg/ingester/client"
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+// ProfilingWrapper is a wrapper around Ingester that adds tenant ID to pprof labels.
+type ProfilingWrapper struct {
+	ing IngesterAPI
+}
+
+func NewIngesterProfilingWrapper(ing IngesterAPI) *ProfilingWrapper {
+	return &ProfilingWrapper{
+		ing: ing,
+	}
+}
+
+type labelNamesAndValuesStream struct {
+	ctx context.Context
+	client.Ingester_LabelNamesAndValuesServer
+}
+
+func (s labelNamesAndValuesStream) Context() context.Context {
+	return s.ctx
+}
+
+type labelValuesCardinalityStream struct {
+	ctx context.Context
+	client.Ingester_LabelValuesCardinalityServer
+}
+
+func (s labelValuesCardinalityStream) Context() context.Context {
+	return s.ctx
+}
+
+type activeSeriesStream struct {
+	ctx context.Context
+	client.Ingester_ActiveSeriesServer
+}
+
+func (s activeSeriesStream) Context() context.Context {
+	return s.ctx
+}
+
+type queryStreamStream struct {
+	ctx context.Context
+	client.Ingester_QueryStreamServer
+}
+
+func (s queryStreamStream) Context() context.Context {
+	return s.ctx
+}
+
+// isTraceSampled checks if the current trace is sampled
+func isTraceSampled(ctx context.Context) bool {
+	return trace.SpanFromContext(ctx).SpanContext().IsSampled()
+}
+
+func (i *ProfilingWrapper) Push(ctx context.Context, request *mimirpb.WriteRequest) (*mimirpb.WriteResponse, error) {
+	if isTraceSampled(ctx) {
+		userID, _ := tenant.TenantID(ctx)
+		labels := pprof.Labels("method", "Ingester.Push", "userID", userID)
+		ctx = pprof.WithLabels(ctx, labels)
+		pprof.SetGoroutineLabels(ctx)
+		defer pprof.SetGoroutineLabels(ctx)
```

Comment:
The `defer` call here, and in the methods below, looks off to me. If we want to recover the original context of the goroutine we should schedule the `defer` call **before** we override the local `ctx`, right?

Reply:
oh you're right. i didn't realize the point of that defer